src/rust/engine/fs/src/lib.rs

```diff
@@ -729,29 +900,137 @@ pub trait VFS<E: Send + Sync + 'static>: Clone + Send + Sync + 'static {
-      round.map(move |paths_and_globs| {
+      round.map(move |single_expansion_results| {
         // Collect distinct new PathStats and PathGlobs
-        for (paths, globs) in paths_and_globs.into_iter() {
-          expansion.outputs.extend(paths.into_iter().map(|p| (p, ())));
-          let completed = &mut expansion.completed;
+        for exp in single_expansion_results {
+          let SingleExpansionResult {
+            sourced_glob: GlobWithSource { path_glob, source },
+            path_stats,
+            globs,
+          } = exp;
+
+          expansion.outputs.extend(path_stats.clone());
+
           expansion
-            .todo
-            .extend(globs.into_iter().filter(|pg| completed.insert(pg.clone())));
+            .completed
+            .entry(path_glob.clone())
+            .or_insert_with(|| GlobExpansionCacheEntry {
+              globs: globs.clone(),
+              matched: if path_stats.is_empty() {
+                GlobMatch::DidNotMatchAnyFiles
+              } else {
+                GlobMatch::SuccessfullyMatchedSomeFiles
+              },
+              sources: vec![],
+            })
+            .sources
+            .push(source);
+
+          // TODO(cosmicexplorer): is cloning these so many times as `GlobSource`s something we can
+          // bypass by using `Arc`s? Profile first.
+          let source_for_children = GlobSource::ParentGlob(path_glob);
+          for child_glob in globs {
+            if expansion.completed.contains_key(&child_glob) {
+              expansion
+                .completed
+                .get_mut(&child_glob)
+                .unwrap()
+                .sources
+                .push(source_for_children.clone());
+            } else {
+              expansion.todo.push(GlobWithSource {
+                path_glob: child_glob,
+                source: source_for_children.clone(),
+              });
+            }
+          }
         }
 
         // If there were any new PathGlobs, continue the expansion.
         if expansion.todo.is_empty() {
-          future::Loop::Break(expansion.outputs)
+          future::Loop::Break(expansion)
         } else {
           future::Loop::Continue(expansion)
         }
       })
-    }).map(|expansion_outputs| {
+    }).and_then(move |final_expansion| {
       // Finally, capture the resulting PathStats from the expansion.
-      expansion_outputs.into_iter().map(|(k, _)| k).collect()
+      let PathGlobsExpansion {
+        outputs,
+        mut completed,
+        ..
+      } = final_expansion;
+
+      let match_results: Vec<_> = outputs.into_iter().collect();
+
+      let PathGlobs {
+        include_entries,
+        exclude,
+        strict_match_behavior,
+      } = path_globs;
+
+      if strict_match_behavior.should_check_initial_glob_matches() {
+        let mut inputs_with_matches: HashSet<GlobParsedSource> = HashSet::new();
+
+        let all_globs: Vec<PathGlob> = completed.keys().rev().map(|pg| pg.clone()).collect();
```

Comment:
Hm, why reverse order here?

I've run out of time to review this very deeply, so I'm going to blindly "shipit" this block for now, heh.

Reply:
I'll leave a descriptive comment. Basically, we're propagating whether a glob matched anything up from the leaf nodes (which are `PathGlob`s which matched some `PathStat`s) back to the sources, and if you do that in the forward direction, the answer will always be no unless one of the root globs is also a leaf (which is feasible).

Reply:
(not necessarily a tree either but the idea is we're doing a reverse topo sort -- we know we will have checked all the children of some leaf by the time we check the leaf because we use the `IndexMap` to preserve the partial order)